### PR TITLE
DM-38498: rewrite QuantumGraph generation

### PR DESCRIPTION
**Pull request by TallJimbo**
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

**Review by andy-slac — COMMENTED** [review]
I only looked at the first four commits and I have to stop for today. Here is a small number of comments, will continue tomorrow.

**Review by TallJimbo — COMMENTED** [review]
I've addressed the first round of review comments (lots of fixup commits that will be squashed later) and squashed a bug discovered by Jenkins (mostly missing exception-handling around registry queries).

**Review by andy-slac — APPROVED** [review]
I checked all remaining commits, looks great, a couple of minor comments.

File: python/lsst/pipe/base/quantum_graph_builder.py [hunk 1]
```diff
@@ -169,6 +167,7 @@ def __init__(
         skip_existing_in: Sequence[str] = (),
         clobber: bool = False,
     ):
+        self.log = getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (TallJimbo) [reply python/lsst/pipe/base/quantum_graph_builder.py:172]: Thanks for the heads-up; I did not know about the issue with logs and pickling/multiprocessing.

I think I will leave this as it is - as you said, multiprocessing is unlikely to be an issue, and I need both `self.log` and `self.metadata` to allow the `timeMethod` decorator to work.

File: python/lsst/pipe/base/quantum_graph_skeleton.py [hunk 2]
```diff
@@ -102,14 +112,16 @@ class PrerequisiteDatasetKey(NamedTuple):
     parent_dataset_type_name: str
     """Name of the dataset type (never a component)."""
 
-    dataset_id: DatasetId
-    """Dataset ID (UUID)."""
+    dataset_id_bytes: bytes
+    """Dataset ID (UUID) as raw bytes."""
```
[the remark this answers is not in the document]
Comment (TallJimbo) [reply python/lsst/pipe/base/quantum_graph_skeleton.py:116]: It's never actually accessed directly, but it is used as part of the (generated) `namedtuple` `__eq__` and `__hash__`.

I guess `int` might be more efficient if that can be boiled down to a native `uint128` on platforms that support it, but I am pretty certain this is not at all a bottleneck - converting the `DataCoordinate` attributes to `tuples` on this commit was _huge_ for performance (several orders of magnitude), but hotspot where the set operations on those occur is in the loop over query results before any prerequisites are added.

**Pull request merged**